src/lib/coda_base/transaction_logic.ml

```diff
@@ -927,6 +1045,58 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
         set ledger receiver_location receiver_account ;
         set ledger source_location source_account ;
         remove_accounts_exn ledger previous_empty_accounts
+    | Create_new_token _, Create_new_token _ ->
+        let fee_payer_account =
+          let balance =
+            Option.value_exn
+              (Balance.add_amount fee_payer_account.balance
+                 (Amount.of_fee constraint_constants.account_creation_fee))
+          in
+          {fee_payer_account with balance}
+        in
+        let%bind source_location =
+          location_of_account' ledger "source" source
+        in
+        let%map source_account =
+          if Account_id.equal fee_payer source then return fee_payer_account
+          else get' ledger "source" source_location
+        in
+        let receiver =
+          User_command.receiver ~next_available_token user_command
+        in
+        set ledger fee_payer_location fee_payer_account ;
+        set ledger source_location
+          { source_account with
+            timing= Option.value ~default:source_account.timing source_timing
+          } ;
+        remove_accounts_exn ledger [receiver] ;
+        (* Restore to the previous [next_available_token]. *)
+        set_next_available_token ledger next_available_token
+    | Create_token_account _, Create_token_account ->
+        let fee_payer_account =
```

Comment:
can the code in the two clauses be factored out? it appears mostly identical, except the last lines.